File: 001_050/Java/A002.java
package A001_A050;

public class A002 {
	ListNode l1 = new ListNode(2);
	ListNode l2 = new ListNode(3);

	public class ListNode {
		int val;
		ListNode next;

		ListNode(int x) {
			val = x;
		}
	}

	public void addNode(ListNode l, int d) {
		ListNode current_node = new ListNode(d);
		while (l.next != null) {
			l = l.next;
		}
		l.next = current_node;
	}

	public void print_list(ListNode l) {
		while (l != null) {
			System.out.print(l.val + " ");
			l = l.next;
		}
		System.out.println("");
	}

	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		ListNode result = new ListNode(0);
		ListNode p = result;
		int carray = 0;
		while ((l1 != null) || (l2 != null) || carray != 0) {
			int sum = (l1 != null ? l1.val : 0) + (l2 != null ? l2.val : 0) + carray;
			carray = sum / 10;
			p.next = new ListNode(sum % 10);
			p = p.next;
			l1 = l1 != null ? l1.next : l1;
			l2 = l2 != null ? l2.next : l2;
		}
		return result.next;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub

		A002 a002 = new A002();
		a002.addNode(a002.l1, 3);
		a002.addNode(a002.l1, 5);
		a002.addNode(a002.l1, 9);
		a002.addNode(a002.l1, 9);
		a002.print_list(a002.l1);

		a002.addNode(a002.l2, 7);
		a002.addNode(a002.l2, 9);
		a002.print_list(a002.l2);

		a002.print_list(a002.addTwoNumbers(a002.l1, a002.l2));

	}

}

File: 001_050/python/A002.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        ptrHead = ListNode(0)
        p = ptrHead
        carry = 0
        while((l1 is not None) or (l2 is not None) or carry != 0):
            sum = (l1.val if (l1 is not None) else 0) + (l2.val if (l2 is not None) else 0) + carry
            carry = int(sum/10)
            p.next = ListNode(sum % 10)
            p = p.next
            l1 = l1.next if l1 is not None else l1
            l2 = l2.next if l2 is not None else l2
        return ptrHead.next

    def print_list(self, l):
        while(l is not None):
            print(l.val)
            l = l.next

l1 = ListNode(2)
temp_1 = ListNode(3)
temp_2 = ListNode(5)
temp_3 = ListNode(9)
temp_4 = ListNode(9)
l1.next = temp_1
temp_1.next = temp_2
temp_2.next = temp_3
temp_3.next = temp_4
#Solution().print_list(l1)

l2 = ListNode(3)
temp_5 = ListNode(7)
temp_6 = ListNode(9)
l2.next = temp_5
temp_5.next = temp_6
#Solution().print_list(l2)

l = Solution().addTwoNumbers(l1, l2)
Solution().print_list(l)


